lightning/src/ln/channelmanager.rs

```diff
@@ -9121,16 +9122,32 @@ macro_rules! create_offer_builder { ($self: ident, $builder: ty) => {
 
 		let nonce = Nonce::from_entropy_source(entropy);
 		let context = OffersContext::InvoiceRequest { nonce };
-		let path = $self.create_blinded_paths_using_absolute_expiry(context, absolute_expiry)
-			.and_then(|paths| paths.into_iter().next().ok_or(()))
-			.map_err(|_| Bolt12SemanticError::MissingPaths)?;
-		let builder = OfferBuilder::deriving_signing_pubkey(node_id, expanded_key, nonce, secp_ctx)
-			.chain_hash($self.chain_hash)
-			.path(path);
+		let builder = match blinded_path {
+			Some(BlindedPathType::Compact) => {
+				let path = $self
+					.create_compact_blinded_paths(context)
+					.and_then(|paths| paths.into_iter().next().ok_or(()))
+					.map_err(|_| Bolt12SemanticError::MissingPaths)?;
 
-		let builder = match absolute_expiry {
-			None => builder,
-			Some(absolute_expiry) => builder.absolute_expiry(absolute_expiry),
+				OfferBuilder::deriving_signing_pubkey(node_id, expanded_key, nonce, secp_ctx)
+					.chain_hash($self.chain_hash)
+					.path(path)
+			}
+
+			Some(BlindedPathType::Full) => {
+				let context = MessageContext::Offers(context);
+				let path = $self
+					.create_blinded_paths(context)
+					.and_then(|paths| paths.into_iter().next().ok_or(()))
```

Comment:
Hmm... @TheBlueMatt I don't think this interface is sufficient to allow someone to specify they want more than one path in the offer, unless we include all the paths returned by the `MessageRouter` instead of just one.

Reply:
What should we have `DefaultMessageRouter` do?

Reply:
I suppose `DefaultMessageRouter` could use _both_ `MessageContex` and `BlindedPathType` -- or which `MessageRouter` method is called, if we revert to that state -- to determine if more than one path should be returned. That way we'd allow users to pass `BlindedPathType::Full` to `create_offer_builder` such that they can create an offer with more than one path (e.g., when an offer doesn't need to be in a QR code).

Reply:
Yeah, seems like either of these approaches should suffice. I'd prefer the two separate method approach, so basically keeping the `MessageRouter` as it was and changing `DefaultMessageRouter::create_compact_blinded_paths` to only return one path.

Reply:
The point was to allow users creating an offer/refund via `ChannelManager` to specify whether they want no blinded paths, blinded paths that can fit in a QR code, or blinded paths where size constraints is not a limitation.
- Passing `None` for `BlindedPathType` adds support for the first
- Passing `Some(BlindedPathType::Compact)` keeps support for the second
- Passing `Some(BlindedPathType::Full)` and selecting all paths returned by `MessageRouter` adds support for the third

Reply:
If we change `MessageRouter` to only have one method, then we need to pass in `BlindedPathType`. But then the `peers` parameter will need to be `Vec<MessageForwardNode>` even when used with `BlindedPathType::Full` instead of `Vec<PublicKey>`. So callers could pass in a `MessageForwardNode` with an scid even though it isn't expected, resulting in the last hop being compact.

Reply:
> So callers could pass in a MessageForwardNode with an scid even though it isn't expected, resulting in the last hop being compact.

Makes sense! I have reintroduced the two function flow back in [pr3246.11](https://github.com/shaavan/rust-lightning/commits/pr3246.11)

Reply:
> I'm confused why this is a concern, if the `MessageRouter` insists on a specific type of blinded path in general, so what? Ideally the `MessageRouter` gets to pick what it wants to do,

Nothing is preventing a `MessageRouter` from doing so. It would just need access to the channels.

It's just kinda ugly for `DefaultMessageRouter` to need to clear `short_channel_id` depending on how it is called. Plus, `ChannelManager` needs to do additional lookups to get this information just to have it discarded. I'm fine either way, but it doesn't optimize the usual case (i.e., use non-compact paths for reply paths).

> though I think we're not allowed to have a compact introduction point in reply paths? Even in that case, though, ISTM we can communicate the restrictions and if the `MessageRouter` is buggy we can just fail.

Yeah, though it doesn't look like we are enforcing this at the moment.